src/Lecture2.hs

```diff
@@ -291,5 +370,21 @@ x + 45 + y
 Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
+
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding = createOptimizedExpr . foldAllLiterals . partition isVar . flattenExpr []
+  where
+    isVar expr = case expr of
+                  Var _ -> True
+                  _     -> False
+    flattenExpr ctn expr = case expr of
+                            Lit _          -> expr : ctn
+                            Var _          -> expr : ctn
+                            Add left right -> flattenExpr ctn left ++ flattenExpr ctn right
+    foldAllLiterals (vars, literals) = foldl (\(vs, Lit res) (Lit val) -> 
+                                                (vs, Lit (res + val))
+                                            ) (vars, Lit 0) literals
```

Comment:
You're using the "Boolean blindness" antipattern here. This happens when you validate values by some Boolean predicate and then assume that the data is correct and use partial functions or non-exhaustive pattern-matching. Instead, it's better to make invalid states unrepresentable:

* https://kowainik.github.io/posts/haskell-mini-patterns

Here you can write a function that converts a single `Expr` to `([String], Int)` — list of variable names and the sum of all literals in a single traversal. Then you can pattern match on this value to produce the final result 🙂